plugin/pkg/utils/cloudprofile.go

```diff
@@ -0,0 +1,143 @@
+	if err != nil {
+		return err
+	}
+	if oldCloudProfileReference.Kind == constants.CloudProfileReferenceKindCloudProfile &&
+		newCloudProfileReference.Kind == constants.CloudProfileReferenceKindNamespacedCloudProfile &&
+		equality.Semantic.DeepEqual(oldCloudProfileReference, newCloudProfileRoot) {
+		return nil
+	}
+
+	return fmt.Errorf("a Seed's CloudProfile may only be changed to a descendant NamespacedCloudProfile, other modifications are currently not supported")
+}
+
+// getRootCloudProfile determines the root CloudProfile from a CloudProfileReference containing any (Namespaced)CloudProfile
+func getRootCloudProfile(namespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, cloudProfile *gardencorev1beta1.CloudProfileReference, namespace string) (*gardencorev1beta1.CloudProfileReference, error) {
+	if cloudProfile == nil {
+		return nil, errors.New("unexpected nil cloudprofile to get root of")
+	}
+	switch cloudProfile.Kind {
+	case constants.CloudProfileReferenceKindCloudProfile:
+		return cloudProfile, nil
+	case constants.CloudProfileReferenceKindNamespacedCloudProfile:
+		cp, err := namespacedCloudProfileLister.NamespacedCloudProfiles(namespace).Get(cloudProfile.Name)
+		if err != nil {
+			return nil, err
+		}
+		return getRootCloudProfile(namespacedCloudProfileLister, &cp.Spec.Parent, namespace)
+	}
+	return nil, fmt.Errorf("unexpected cloudprofile kind %s", cloudProfile.Kind)
+}
+
+// BuildCloudProfileReference determines the CloudProfile of a Shoot to use
+// depending on the availability of cloudProfile and cloudProfileName.
+func BuildCloudProfileReference(shoot *core.Shoot) *gardencorev1beta1.CloudProfileReference {
+	if shoot == nil {
+		return nil
+	}
+	if shoot.Spec.CloudProfile != nil {
+		cloudProfileV1Beta1 := &gardencorev1beta1.CloudProfileReference{}
+		if err := api.Scheme.Convert(shoot.Spec.CloudProfile, cloudProfileV1Beta1, nil); err != nil {
+			return nil
+		}
+		if len(cloudProfileV1Beta1.Kind) == 0 {
+			cloudProfileV1Beta1.Kind = constants.CloudProfileReferenceKindCloudProfile
+		}
+		return cloudProfileV1Beta1
+	}
+	if len(ptr.Deref(shoot.Spec.CloudProfileName, "")) > 0 {
+		return &gardencorev1beta1.CloudProfileReference{
+			Name: *shoot.Spec.CloudProfileName,
+			Kind: constants.CloudProfileReferenceKindCloudProfile,
+		}
+	}
+	return nil
+}
+
+// SyncCloudProfileFields handles the coexistence of a Shoot Spec's cloudProfileName and cloudProfile
+// by making sure both fields are synced correctly and appropriate fallback cases are handled.
+func SyncCloudProfileFields(newShoot *core.Shoot) {
+	// clear cloudProfile if namespacedCloudProfile is provided but feature toggle is disabled
+	if newShoot.Spec.CloudProfile != nil && newShoot.Spec.CloudProfile.Kind == constants.CloudProfileReferenceKindNamespacedCloudProfile && !utilfeature.DefaultFeatureGate.Enabled(features.UseNamespacedCloudProfile) {
```

Comment:
This will break all shoots which have been created with a NamespacedCloudprofile, when the feature gate is disabled afterwards. I thought that an already set NamespacedCloudProfile will remain as is in this case?

Reply:
see other comment, added test case and corrected logic.